### DISCUSSION TITLE
Provide a context object

### DISCUSSION THREAD

**Discussion by SebastianStehle** [microsoft/semantic-kernel#5345] [Ideas]
We are not using semantic kernel yet and rely on node so far, e.g. langchain. But in my heart I am  .NET developer.

What I am really missing here is some kind of context object that is available through the whole flow. Something like HttpContext with access to the current user and other application data. I know that we could probably rely on AsyncLocal, but this should not be necessary in my opinion.

Many, if not all, of our RAG scenarios need a permission system to restrict the data.

**Comment by stephentoub** [comment]
The Kernel object is passed down through to everything involved in the operation, and it has a Data dictionary into which you can store any state you want. It's there explicitly so you can provide whatever arbitrary data you want and have that flow into the various connectors, functions, etc.

**Reply by SebastianStehle** [reply]
Awesome. I haven't thought about that one. I have assumed it would be a singleton. Thanks :)

**Reply by stephentoub** [reply]
No problem. Kernel _can_ be a singleton if you really want, but it needn't be. In fact, the built-in AddKernel method that adds Kernel to a dependency injection IServiceCollection does so as transient, such that you get a new Kernel each time you resolve one. It's intended to be a lightweight thing you can construct, modify for a given use, and then throw away. Such modifications can include adding more plugins, adding filters, adding data into the Data dictionary, etc.
